### PR TITLE
Multi source example

### PR DESCRIPTION
**Pull request by sz144**
### Description
A new example for demonstrating multi-source domain adaptation.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] In-line docstrings updated.
- [ ] [Source for documentation at `docs`](https://github.com/pykale/pykale/tree/main/docs/source) manually updated for new API.

### CODE REVIEW

File: examples/multisource_adapt/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+`python main.py --cfg configs/Office2A-MFSAN.yaml --gpus 0`
+
+### 3. Related `kale` API
+
+`kale.embed.image_cnn`: Extract features from small-size (32x32) images using CNN.
+
+`kale.loaddata.digits_access`: Data loaders for digits datasets.
+
+`kale.loaddata.image_access`: Data loaders for general APIs for image datasets.
```
Comment (haipinglu) [inline examples/multisource_adapt/README.md:26]: Check, hard to parse

File: examples/multisource_adapt/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+This example demonstrates multi-source domain adaptation methods with application in image object detection/recognition.
+
+### 2. Usage
+
+* Datasets:
+  * Digits (10 classes), 4 domains: MNIST, Modified MNIST (MNISTM), UPSP, SVHN
+  * Office-31 (31 classes), 3 domains: Amazon (A), DSLR (D), Webcam (W)
+  * Office-Caltech (10 classes), 4 domains: Amazon (A), DSLR (D), Webcam (W), Caltech (C)
+* Algorithms: M3SDA, MFSAN
+* Example: Caltech, DSLR, and Webcam (source) to Amazon (target) using M3SDA and MFSAN
```
Comment (haipinglu) [inline examples/multisource_adapt/README.md:14]: (source) --> (three sources)

File: examples/multisource_adapt/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+`kale.loaddata.image_access`: Data loaders for general APIs for image datasets.
+
+`kale.loaddata.mnistm`: Data loader for the [modified MNIST data](https://github.com/zumpchke/keras_mnistm).
+
+`kale.loaddata.multi_domain`: Construct the dataset for (multiple) source and target domains.
+
+`kale.loaddata.usps`: Data loader for the [USPS data](https://git-disl.github.io/GTDLBench/datasets/usps_dataset/).
+
+`kale.pipeline.multi_domain_adapter`: Multi-source Domain adaptation pipelines for image classification.
```
Comment (haipinglu) [inline examples/multisource_adapt/README.md:34]: why cap "Domain"?

File: examples/multisource_adapt/config.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+"""
+Default configurations for multi-source domain adapation
+"""
+import os
+
+from yacs.config import CfgNode as CN
+
+# import os
```
Comment (haipinglu) [inline examples/multisource_adapt/config.py:8]: remove? it is above

File: examples/multisource_adapt/config.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+_C = CN()
+
+# -----------------------------------------------------------------------------
+# Dataset
+# -----------------------------------------------------------------------------
+_C.DATASET = CN()
+_C.DATASET.ROOT = "../data"
+_C.DATASET.NAME = "digits"  # dset choices=['office', 'image-clef', 'office-home']
+# _C.DATASET.SOURCE = ["cartoon", "art_painting", "photo"]
```
Comment (haipinglu) [inline examples/multisource_adapt/config.py:22]: no default source? It seems the default configs won't run (w/o specifying a default source)

File: examples/multisource_adapt/config.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+_C.DATASET.NUM_REPEAT = 10  # 10
+_C.DATASET.NUM_CHANNELS = 3
+_C.DATASET.DIMENSION = 784
+_C.DATASET.WEIGHT_TYPE = "natural"
+_C.DATASET.SIZE_TYPE = "source"
+# ---------------------------------------------------------------------------- #
+# Solver
+# ---------------------------------------------------------------------------- #
+_C.SOLVER = CN()
+_C.SOLVER.SEED = 2020
```
Comment (haipinglu) [inline examples/multisource_adapt/config.py:35]: 2021 与时俱进

File: examples/multisource_adapt/config.py (hunk 5 above)
Comment (haipinglu) [reply examples/multisource_adapt/config.py:22]: do you imply the sources?

Comment (haipinglu) [reply examples/multisource_adapt/config.py:22]: what if I just want to use two sources (faster, e.g. for notebooks) rather than all three?

File: examples/multisource_adapt/configs/Office2A-M3SDA.yaml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+DAN:
+  METHOD: "M3SDA"
+DATASET:
+  NAME: "office_caltech"
+  NUM_CLASSES: 10
+  ROOT: "../data/office"
+  TARGET: "amazon"
```
Comment (haipinglu) [inline examples/multisource_adapt/configs/Office2A-M3SDA.yaml:7]: I guess you want to say that the sources are implied to be the other three domains, add a comment maybe e.g. after target? Do you support 2 sources?

File: examples/multisource_adapt/configs/tutorial.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
+DAN:
+  METHOD: "M3SDA"
+DATASET:
+  NAME: "OFFICE_CALTECH"
+  NUM_CLASSES: 10
+  ROOT: "../data/Office"
+  TARGET: "amazon"
```
Comment (haipinglu) [inline examples/multisource_adapt/configs/tutorial.yaml:7]: for tutorial, you should use the min setting, i.e. using two sources.

File: examples/multisource_adapt/main.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import os
+import sys
+
+import pytorch_lightning as pl
+from config import get_cfg_defaults
+from model import get_model
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.loggers import TensorBoardLogger
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
```
Comment (haipinglu) [inline examples/multisource_adapt/main.py:17]: is this still necessary? we stopped using this long time ago. See digits example.

File: examples/multisource_adapt/model.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+# Author: Shuo Zhou
+# Initial Date: 09.09.2021
+
+import os
+import sys
+from copy import deepcopy
+
+import torch
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "../..")))
```
Comment (haipinglu) [inline examples/multisource_adapt/model.py:13]: I think no more need.

File: examples/multisource_adapt/model.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                    "momentum": cfg.SOLVER.MOMENTUM,
+                    "weight_decay": cfg.SOLVER.WEIGHT_DECAY,
+                    "nesterov": cfg.SOLVER.NESTEROV,
+                },
+            },
+        },
+        "data_params": {
+            "dataset_group": cfg.DATASET.NAME,
+            # "dataset_name": cfg.DATASET.SOURCE + '2' + cfg.DATASET.TARGET,
+            "dataset_name": "rest2" + cfg.DATASET.TARGET,
```
Comment (haipinglu) [inline examples/multisource_adapt/model.py:47]: I think the name is not precise by omitting sources. Consider `rest` to target

**Review by haipinglu — COMMENTED** [review]
Great to see the more generic image access. Remove digits access in future (in a separate PR)? If yes, create an issue or put a card in project.

Some minor issues remain. Discuss with me if unsure. Otherwise, good to go once fixed.

File: examples/multisource_adapt/README.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+* Algorithms: M3SDA, MFSAN
+* Example: Caltech, DSLR, and Webcam (source) to Amazon (target) using M3SDA and MFSAN
+
+`python main.py --cfg configs/Office2A-M3SDA.yaml --gpus 0`
+
+`python main.py --cfg configs/Office2A-MFSAN.yaml --gpus 0`
+
+### 3. Related `kale` API
+
+`kale.embed.image_cnn`: Extract features from small-size (32x32) images using CNN.
```
Comment (haipinglu) [inline examples/multisource_adapt/README.md:22]: is office data still `small-size (32x32)`?

Comment (sz144) [reply examples/multisource_adapt/README.md:22]: No. This example also supports digits data.

File: examples/multisource_adapt/config.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+_C = CN()
+
+# -----------------------------------------------------------------------------
+# Dataset
+# -----------------------------------------------------------------------------
+_C.DATASET = CN()
+_C.DATASET.ROOT = "../data"
+_C.DATASET.NAME = "digits"  # choices=['office', 'digits', 'office_caltech', 'office31']
+_C.DATASET.SOURCE = None
```
Comment (haipinglu) [inline examples/multisource_adapt/config.py:20]: Explain what None means and how to change it please

File: kale/loaddata/image_access.py [hunk 14]
```diff
@@ -105,3 +108,40 @@ def __init__(self, root, **kwargs):
...
+class ImageAccess:
+    @staticmethod
+    def get_multi_domain_images(image_set_name: str, data_path: str, sub_domain_set=None, **kwargs):
+        """Get multi-domain images as a dataset from the given data path.
+
+        Args:
+            image_set_name (str): name of image dataset
+            data_path (str): path to the image dataset
+            sub_domain_set (list, optional): A list of domain names, which should be a subset of domains in image folders.
+                Defaults to None.
```
Comment (haipinglu) [inline kale/loaddata/image_access.py:122]: Explain what None means

File: tests/loaddata/test_office_access.py [hunk 15]
```diff
@@ -1,11 +1,13 @@
 from numpy import testing
```
Comment (haipinglu) [inline tests/loaddata/test_office_access.py:1]: since you've renamed office_access to image_access, should this module be renamed as well for consistency?

**Review by haipinglu — COMMENTED** [review]
Two more minor comments to address. After which, it will be good to go.

File: tests/loaddata/test_office_access.py (hunk 15 above)
Comment (sz144) [reply tests/loaddata/test_office_access.py:1]: Yes. I will rename it in the next PR. The changes cannot be displayed if I rename it now.

**Review by haipinglu — APPROVED** [review]
